sdk/nodejs/policy/policy.ts

```diff
@@ -121,42 +121,29 @@ export interface ResourceValidationArgs {
 
 /**
  * A helper function that returns a strongly-typed resource validation function.
- * @param typeFilter A type guard used to determine if the args are an instance of the resource.
+ * @param resourceClass A resource class used to filter this check to only resources of the specified class and
+ * determine the appropriate strongly-typed `TArg` type to use for the resource.
  * @param validate A callback function that validates if the resource definition violates a policy.
  */
-export function validateTypedResource<TResource extends Resource>(
-    typeFilter: (o: any) => o is TResource,
+export function validateTypedResource<TResource extends Resource, TArgs>(
+    resourceClass: { new(name: string, args: TArgs, ...rest: any[]): TResource },
```

Comment:
When you pass `aws.s3.Bucket` here, the type of its args (`TArgs`) is inferred from the constructor parameters. This relies on the pattern we generally follow for our resource types, where the first parameter is the name of the resource and the second parameter is it's args bag.

Reply:
So if we eve deviate from this pattern we'd be in trouble here? I wondering if theres anywhere we do deviate / will this also work in other languages?

Reply:
> So if we ever deviate from this pattern we'd be in trouble here?

If we deviate from the pattern, the resource won't be able to make use of the strongly-typed helper. You'd have to use a regular untyped validation function. It does make me a little uneasy, but this should work and provide the nicer experience we want for the vast majority of the resources we provide (I suspect ~99%), and for the cases where it can't be used (can we can't just go fix the resource to follow the pattern), you can still easily workaround using a regular untyped validation function.

> I wondering if theres anywhere we do deviate / will this also work in other languages?

I will look. I suspect that if there are any places where we deviate, it would be in our high-level libraries like AWSX that have resources that inherit from `ComponentResource` rather than `CustomResource`. There's a separate issue (#140) with `ComponentResource` in that it doesn't use the `__pulumiType` pattern we're using, but I will fix that subsequently.

Reply:
As for other languages, I still want to do some prototyping, but I don't think we'll be able to automatically infer the args type as nicely as we can with TypeScript.

Reply:
I took a look at resources in `awsx` and `eks`... There are a handful of resources that extend `ComponentResource` that have extra parameters that don't work with the pattern:

```typescript
export class StepScalingPolicy extends pulumi.ComponentResource {
    constructor(name: string, group: AutoScalingGroup,
                args: StepScalingPolicyArgs, opts: pulumi.ComponentResourceOptions = {}) {}
}

export class Subnet extends pulumi.ComponentResource {
    constructor(name: string, vpc: x.ec2.Vpc, args: SubnetArgs, opts?: pulumi.ComponentResourceOptions);
    constructor(name: string, vpc: x.ec2.Vpc, args: ExistingSubnetArgs, opts?: pulumi.ComponentResourceOptions);
    constructor(name: string, vpc: x.ec2.Vpc, args: SubnetArgs | ExistingSubnetArgs, opts: pulumi.ComponentResourceOptions = {}) {}
}

export class EgressSecurityGroupRule extends SecurityGroupRule {
    constructor(name: string, securityGroup: x.ec2.SecurityGroup,
                args: SimpleSecurityGroupRuleArgs | EgressSecurityGroupRuleArgs,
                opts?: pulumi.ComponentResourceOptions) {}
}

export class IngressSecurityGroupRule extends SecurityGroupRule {
    constructor(name: string, securityGroup: x.ec2.SecurityGroup,
                args: SimpleSecurityGroupRuleArgs | IngressSecurityGroupRuleArgs,
                opts?: pulumi.ComponentResourceOptions) {}
}

export class LifecyclePolicy extends aws.ecr.LifecyclePolicy {
    constructor(name: string, repository: aws.ecr.Repository, args?: LifecyclePolicyArgs, opts: pulumi.ComponentResourceOptions = {}) {}
}

export class ListenerRule extends pulumi.ComponentResource {
    constructor(name: string, listener: x.lb.Listener,
                args: ListenerRuleArgs, opts: pulumi.ComponentResourceOptions = {}) {
}

export class TargetGroupAttachment extends pulumi.ComponentResource {
    constructor(name: string, targetGroup: mod.TargetGroup, args: mod.LoadBalancerTarget, opts: pulumi.ComponentResourceOptions = {}) {
}
```

This is a pretty small number compared to the vast number of resources from TF and Kubernetes that follow the pattern. There may be some room for another strongly-typed helper that'd fit in-between the low-level untyped validation method and the constructor pattern-based `validateTypedResource` helper, that could be used for such types. Will think about this as part of addressing #140.

In eks, there was a dynamic resource, but we already couldn't filter those because [they all have the same type](https://github.com/pulumi/pulumi/blob/137fd54f1cdd5ff901ddba0f281039bf3b9a5e01/sdk/nodejs/dynamic/index.ts#L189) (`"pulumi-nodejs:dynamic:Resource"`). We should think about whether we want to make it easy to write policies for dynamic resources in a strongly-typed way separately.

```
export class VpcCni extends pulumi.dynamic.Resource {
    constructor(name: string, kubeconfig: pulumi.Input<any>, args?: VpcCniOptions, opts?: pulumi.CustomResourceOptions) {}
}
```

Reply:
Is this something we should call out in docs as a limitation? 🤔